include/deal.II/matrix_free/tensor_product_kernels.h

```diff
@@ -3292,38 +3372,46 @@ namespace internal
       poly[i].values_of_array(point, 1, &shapes[i][0]);
 
     // Implement the transpose of the function above
+    std::array<Number2, 3> test_grads_value;
     for (int i2 = 0, i = 0; i2 < (dim > 2 ? n_shapes : 1); ++i2)
       {
-        const Number2 test_value_z =
+        // test grad x
+        test_grads_value[0] =
+          dim > 2 ? gradient[0] * shapes[i2][0][2] : gradient[0];
+        // test grad y
+        test_grads_value[1] = dim > 2 ? gradient[1] * shapes[i2][0][2] :
+                                        (dim > 1 ? gradient[1] : Number2());
+        // test value z
+        test_grads_value[2] =
           dim > 2 ?
             (value * shapes[i2][0][2] + gradient[2] * shapes[i2][1][2]) :
             value;
-        const Number2 test_grad_x =
-          dim > 2 ? gradient[0] * shapes[i2][0][2] : gradient[0];
-        const Number2 test_grad_y = dim > 2 ?
-                                      gradient[1] * shapes[i2][0][2] :
-                                      (dim > 1 ? gradient[1] : Number2());
-        for (int i1 = 0; i1 < (dim > 1 ? n_shapes : 1); ++i1)
-          {
-            const Number2 test_value_y = dim > 1 ?
-                                           (test_value_z * shapes[i1][0][1] +
-                                            test_grad_y * shapes[i1][1][1]) :
-                                           test_value_z;
-            const Number2 test_grad_xy =
-              dim > 1 ? test_grad_x * shapes[i1][0][1] : test_grad_x;
-            if (renumber.empty())
-              for (int i0 = 0; i0 < n_shapes; ++i0, ++i)
-                values[i] += shapes[i0][0][0] * test_value_y +
-                             shapes[i0][1][0] * test_grad_xy;
-            else
-              for (int i0 = 0; i0 < n_shapes; ++i0, ++i)
-                values[renumber[i]] += shapes[i0][0][0] * test_value_y +
-                                       shapes[i0][1][0] * test_grad_xy;
-          }
+
+        // Generate separate code with known loop bounds for the most common
+        // cases
+        if (n_shapes == 2)
```

Comment:
Instead of spelling out all the cases, you could also write:

```diff
diff --git a/include/deal.II/matrix_free/tensor_product_kernels.h b/include/deal.II/matrix_free/tensor_product_kernels.h
index e3a5c35..420b996 100644
--- a/include/deal.II/matrix_free/tensor_product_kernels.h
+++ b/include/deal.II/matrix_free/tensor_product_kernels.h
@@ -3296,7 +3296,7 @@ namespace internal
 
 
 
-  template <int dim, int length, typename Number2, typename Number>
+  template <int dim, typename Number2, typename Number, int length = 2>
   inline DEAL_II_ALWAYS_INLINE void
   do_test_xy(AlignedVector<Number2> &                    values,
              const std::vector<unsigned int> &           renumber,
@@ -3307,7 +3307,22 @@ namespace internal
              const int                                   n_shapes_runtime,
              int &                                       i)
   {
-    const int n_shapes = length > 0 ? length : n_shapes_runtime;
+    constexpr int max_length = 6;
+
+    if (n_shapes_runtime != length)
+      return do_test_xy<dim,
+                        Number2,
+                        Number,
+                        std::min(max_length + 1, length + 1)>(values,
+                                                              renumber,
+                                                              shapes,
+                                                              test_grad_x,
+                                                              test_grad_y,
+                                                              test_value_z,
+                                                              n_shapes_runtime,
+                                                              i);
+
+    const int n_shapes = length > max_length ? length : n_shapes_runtime;
     for (int i1 = 0; i1 < (dim > 1 ? n_shapes : 1); ++i1)
       {
         const Number2 test_value_y =
@@ -3377,60 +3392,14 @@ namespace internal
                                       gradient[1] * shapes[i2][0][2] :
                                       (dim > 1 ? gradient[1] : Number2());
 
-        if (n_shapes == 2)
-          do_test_xy<dim, 2, Number2, Number>(values,
-                                              renumber,
-                                              shapes,
-                                              test_grad_x,
-                                              test_grad_y,
-                                              test_value_z,
-                                              n_shapes,
-                                              i);
-        else if (n_shapes == 3)
-          do_test_xy<dim, 3, Number2, Number>(values,
-                                              renumber,
-                                              shapes,
-                                              test_grad_x,
-                                              test_grad_y,
-                                              test_value_z,
-                                              n_shapes,
-                                              i);
-        else if (n_shapes == 4)
-          do_test_xy<dim, 4, Number2, Number>(values,
-                                              renumber,
-                                              shapes,
-                                              test_grad_x,
-                                              test_grad_y,
-                                              test_value_z,
-                                              n_shapes,
-                                              i);
-        else if (n_shapes == 5)
-          do_test_xy<dim, 5, Number2, Number>(values,
-                                              renumber,
-                                              shapes,
-                                              test_grad_x,
-                                              test_grad_y,
-                                              test_value_z,
-                                              n_shapes,
-                                              i);
-        else if (n_shapes == 6)
-          do_test_xy<dim, 6, Number2, Number>(values,
-                                              renumber,
-                                              shapes,
-                                              test_grad_x,
-                                              test_grad_y,
-                                              test_value_z,
-                                              n_shapes,
-                                              i);
-        else
-          do_test_xy<dim, -1, Number2, Number>(values,
-                                               renumber,
-                                               shapes,
-                                               test_grad_x,
-                                               test_grad_y,
-                                               test_value_z,
-                                               n_shapes,
-                                               i);
+        do_test_xy<dim, Number2, Number>(values,
+                                         renumber,
+                                         shapes,
+                                         test_grad_x,
+                                         test_grad_y,
+                                         test_value_z,
+                                         n_shapes,
+                                         i);
       }
   }
 
```

Reply:
This is based on my suggestion, so let me explain. If you call this recursively, then the compiler generates a function prologue/epilogue for every step that is not inlined. Given that we should assess the suggestion of @drwells at https://github.com/dealii/dealii/pull/14972#discussion_r1148656523, I would like to ensure that performance is not affected. The current code is based on the experience of `FEEvaluationImpl`, where the effect of stepping into another function to arrive at the target degree is quite visible especially in 2D (around 10% of the instructions, which is pretty bad). Note that this is an inner loop component that does less work than `FEEvaluation` in 2D in all cases, so picking a good strategy is vital. It might be worth measuring for some representative cases, but I would rank the ability to avoid `always_inline` higher than repeating these steps, especially if we manage to fit the call in one or two lines. :wink: